lib/imap.php

```diff
@@ -52,25 +52,29 @@ public function checkPassword($uid, $password) {
 			$uid = str_replace("%40","@",$uid);
 		}
 
-                $result = OC_DB::executeAudited(
-                        'SELECT `userid` FROM `*PREFIX*preferences` WHERE `appid`=? AND `configkey`=? AND `configvalue`=?',
-                        array('settings','email',$uid)
-                );
+		$query = \OC::$server->getDatabaseConnection()->getQueryBuilder();
+		$query->select('userid')
+			->from('preferences')
+			->where($query->expr()->eq('appid', $query->createNamedParameter('settings')))
+			->andWhere($query->expr()->eq('configkey', $query->createNamedParameter('email')))
+			->andWhere($query->expr()->eq('configvalue', $query->createNamedParameter($uid)));
+		$result = $query->execute();
 
-		$users = array();
-		while ($row = $result->fetchRow()) {
+		$users = [];
+		while ($row = $result->fetch()) {
 			$users[] = $row['userid'];
 		}
+		$result->closeCursor();
 
 		if(count($users) === 1) {
 			$username = $uid;
 			$uid = $users[0];
- 		// Check if we only want logins from ONE domain and strip the domain part from UID		
-		}elseif($this->domain != '') {
+ 		// Check if we only want logins from ONE domain and strip the domain part from UID
+		}elseif($this->domain !== '') {
```

Comment:
this code is really strange, when there is not exactly one user, the database result is being ignored. Is this intended?

Reply:
hmm, maybe @kosli can help us here?

Reply:
or maybe @pierreozoux ?

Reply:
@nickvergessen the database returns as many users as there are with the same email address (if this emailaddress is entered in the login form), so this should be zero or one normally, shouldn't it?
It makes it possible to login for normal users with their email address even though in config.php for imap backend another domain is specified. This code as it is written now wouldn't work if there are multiple users with the same email address anyway. If we would change it to `if(count($users) >= 1)` it would always try to login the first user with the given email address... How does Nextcloud's own user backend handle this case?  Shouldn't it be restricted that only one user can have a specific email address?

Reply:
If there are more than 2 users you can not login with email:
https://github.com/nextcloud/server/blob/43d6ae7476bbf4d08991b405a63a0f8dbc2ac25a/core/Controller/LoginController.php#L311-L312

Reply:
I don't suggest to use `if(count($users) >= 1)`, I just wanted to know if that is intended behaviour. Also as per the code I liked above, the imap user backend should not need to do this. It will be done by the user manager. So actually this method (`checkPassword`) will be called twice already for email alike logins, so the user_external app does not need to check this.

Reply:
I commited accordingly, if you still want to keep this, just rebase and drop the commit

Reply:
okay, thank you for clarifying! I'm fine with removing it; login with email for a normal user with another domain still works...